sfcasts/centralize-logic.md

```diff
@@ -1,25 +1,21 @@
 
-Also, we need a way to get access to the parameters. We could inject the whole `ParameterBagInterface` service from which we can access any parameters, but since we only need `storeId` one - let's inject it directly.
+In our constructor, add: `private readonly string $storeId,`. Above, add the PHP attribute with `#[Autowire('%env(LEMON_SQUEEZY_STORE_ID)%')]`. And finally, replace every instance of `$this->getParameter()` with `$this->storeId`. I only see it once here, so that's pretty easy.
 
-In the constructor, add: `private readonly string $storeId,`. Above, add PHP attribute: `#[Autowire('%env(LEMON_SQUEEZY_STORE_ID)%')]`. Now replace `$this->getParameter()` with the `$this->storeId` in all places.
+Now, back in `OrderController::checkout()`, let's get rid of these unused dependencies and inject `LemonSqueezyApi $lsApi` instead. Below, *use* the service with `$lsCheckoutUrl = $lsApi->createCheckoutUrl();`.
 
-Now back to `OrderController::checkout()` - drop unused dependencies and inject `LemonSqueezyApi $lsApi` instead. Below, use the service: `$lsCheckoutUrl = $lsApi->createCheckoutUrl();`.
+Testing time! Let's make sure we can still checkout. On our site, reload, select "Classic Lemonade", add one to the cart, and click "Checkout with LemonSqueezy". *Yes*! We're on the LemonSqueezy checkout page and everything looks great!
 
-Make sure you can still checkout… Yes, we’re on the LemonSqueezy checkout page, everything looks good.
+Okay, now that we know the checkout's working, can we make `$lsStoreUrl` and the `success()` method dynamic? *Indeed*! And LemonSqueezy has an API endpoint for *just* such an occasion! In the API docs, find the "Retrieve a store" endpoint... and check out the example response on the right here. It *looks* like we can read the URL from `attributes`, so, back in our code, in `LemonSqueezyApi`, create a new public method. Call it `retrieveStoreUrl()`, and have it return a `string`. Inside, add `$response = $this->client->request(Request::METHOD_GET, 'stores/' . $this->storeId)`. Below, say `$lsStore = $response->toArray()` and finally, `return $lsStore['data']['attributes']['url']`.
 
-Now, can we make that `$lsStoreUrl` in `success()` URL dynamic? We can! LemonSqueezy has an API endpoint to fetch that store URL. In the API docs, find the [Retrieve a store](https://docs.lemonsqueezy.com/api/stores/retrieve-store) endpoint. If you will take a look at the example response - we can read the URL from `attributes`.
+Back in the `success()` method, inject `LemonSqueezyApi $lsApi,` and replace this hard-coded URL with `$lsStoreUrl = $lsApi->retrieveStoreUrl()`. Time for another test! Back on our site, pick one of our delicious lemonades - I'll choose apple this time - and add it to the cart. On the cart page, click the "Checkout" button again and... tada! It still works!
```

Comment:
@Berealady9 I think after the `click the "Checkout" button again` we need to say something like:

```suggestion
Back in the `success()` method, inject `LemonSqueezyApi $lsApi,` and replace this hard-coded URL with `$lsStoreUrl = $lsApi->retrieveStoreUrl()`. Time for another test! Back on our site, pick one of our delicious lemonades - I'll choose apple this time - and add it to the cart. On the cart page, click the "Checkout" button again, fill in card credentials and billing address, press Pay, and... finally, in the successful modal, press "Continue" and... tada! Here's our flash message, it still works!
```

The main idea is to see that flash message, not only see the LS checkout page.